File: packages/rspec/lib/kensho/rspec/state.rb
# frozen_string_literal: true

# Per-example mutable state used by the public helper API.
#
# The formatter sets the current scratch on each example_started and clears
# it on example_passed/failed/pending. While an example is running, the
# Kensho.step / Kensho.attach / Kensho.label / Kensho.link helpers mutate
# the scratch directly so the formatter can pick up the data when it
# finalizes the case.
#
# RSpec runs examples sequentially within a worker by default; we still use
# Thread.current so threaded fixtures (rare) don't race.

module Kensho
  module RSpec
    class CaseScratch
      attr_accessor :case_id, :example_id, :started_at_ms,
                    :steps, :step_stack, :attachments, :logs, :labels, :links

      def initialize(case_id:, example_id:, started_at_ms:)
        @case_id = case_id
        @example_id = example_id
        @started_at_ms = started_at_ms
        @steps = []
        @step_stack = []
        @attachments = []
        @logs = []
        @labels = {}
        @links = []
      end
    end

    module State
      THREAD_KEY = :__kensho_rspec_scratch__
      GROUP_STACK_KEY = :__kensho_rspec_group_stack__

      def self.current
        Thread.current[THREAD_KEY]
      end

      def self.current=(scratch)
        Thread.current[THREAD_KEY] = scratch
      end

      def self.push_group(group)
        (Thread.current[GROUP_STACK_KEY] ||= []) << group
      end

      def self.pop_group
        stack = Thread.current[GROUP_STACK_KEY]
        stack&.pop
      end

      def self.current_group
        stack = Thread.current[GROUP_STACK_KEY]
        stack ? stack.last : nil
      end

      class << self
        attr_accessor :formatter
      end
    end
  end
end

File: packages/pytest/src/kensho_pytest/_state.py
"""Per-test mutable state used by helper APIs.

The plugin sets ``CURRENT_CASE`` at the start of each test's setup phase
and clears it at the end of teardown. While a test is running, the
:func:`kensho_pytest.step`, :func:`kensho_pytest.attach`,
:func:`kensho_pytest.label`, and :func:`kensho_pytest.link` helpers mutate
this object directly so the plugin can pick up the data when it finalizes
the case.
"""

from __future__ import annotations

import threading
from typing import Any, Dict, List, Optional


class CaseScratch:
    """Mutable per-test scratch space.

    The plugin owns one of these per running test. Helpers append to the
    public lists; the plugin reads them at ``logfinish`` time and merges
    them into the final case JSON.
    """

    __slots__ = (
        "case_id",
        "nodeid",
        "started_at_ms",
        "steps",
        "step_stack",
        "attachments",
        "logs",
        "labels",
        "links",
    )

    def __init__(self, case_id: str, nodeid: str, started_at_ms: float) -> None:
        self.case_id: str = case_id
        self.nodeid: str = nodeid
        self.started_at_ms: float = started_at_ms
        # Top-level steps (children nest via the step stack).
        self.steps: List[Dict[str, Any]] = []
        # Stack of currently-open step dicts; the innermost is .top.
        self.step_stack: List[Dict[str, Any]] = []
        self.attachments: List[Dict[str, Any]] = []
        self.logs: List[Dict[str, Any]] = []
        self.labels: Dict[str, str] = {}
        self.links: List[Dict[str, str]] = []


# Threading note: pytest-xdist runs each worker in its own process, so a
# plain module-level slot is fine for the common case. We still guard with
# a lock so threaded fixtures (rare but possible) don't race the helpers.
_lock = threading.Lock()
_current: Optional[CaseScratch] = None

# Set by ``pytest_configure`` so the public helper API can find the running
# plugin without round-tripping through pytest's config singleton.
PLUGIN: Optional[Any] = None  # type: ignore[assignment]


def set_current(scratch: Optional[CaseScratch]) -> None:
    global _current
    with _lock:
        _current = scratch


def get_current() -> Optional[CaseScratch]:
    return _current


def set_plugin(plugin: Optional[Any]) -> None:
    global PLUGIN
    PLUGIN = plugin
